src/lf_msqueue.ml

```diff
@@ -0,0 +1,100 @@
+(*
+ * Copyright (c) 2015, Théo Laurent <theo.laurent@ens.fr>
+ * Copyright (c) 2015, KC Sivaramakrishnan <sk826@cl.cam.ac.uk>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+(* Michael-Scott queue *)
+
+(* TODO KC: Replace with concurrent lock free bag --
+ * http://dl.acm.org/citation.cfm?id=1989550 *)
+
+type 'a node =
+  | Nil
+  | Next of 'a * 'a node Atomic.t
+
+type 'a t =
+  { head : 'a node Atomic.t ;
+    tail : 'a node Atomic.t }
+
+let create () =
+  let head = (Next (Obj.magic (), Atomic.make Nil)) in
+  { head = Atomic.make head ; tail = Atomic.make head }
+
+let is_empty q =
+  match Atomic.get q.head with
+  | Nil -> failwith "MSQueue.is_empty: impossible"
+  | Next (_,x) ->
+     ( match Atomic.get x with
+       | Nil -> true
+       | _ -> false )
+
+let pop q =
+  let b = Backoff.create () in
+  let rec loop () =
+    let s = Atomic.get q.head in
+    let nhead = match s with
+      | Nil -> failwith "MSQueue.pop: impossible"
+      | Next (_, x) -> Atomic.get x
+    in match nhead with
+     | Nil -> None
+     | Next (v, _) when Atomic.compare_and_set q.head s nhead -> Some v
+     | _ -> ( Backoff.once b ; loop () )
+  in loop ()
+
+let push q v =
+  let rec find_tail_and_enq curr_end node =
+    if Atomic.compare_and_set curr_end Nil node then ()
+    else match Atomic.get curr_end with
+         | Nil -> find_tail_and_enq curr_end node
+         | Next (_, n) -> find_tail_and_enq n node
+  in
+  let newnode = Next (v, Atomic.make Nil) in
+  let tail = Atomic.get q.tail in
+  match tail with
+  | Nil         -> failwith "HW_MSQueue.push: impossible"
+  | Next (_, n) -> begin
+      find_tail_and_enq n newnode;
+      ignore (Atomic.compare_and_set q.tail tail newnode)
+  end
+
+let clean_until q f =
```

Comment:
I'm suspicious that this function may not linearize if called from more than 1 thread. I'll try to create a test that breaks it. 

Assume queue with the following values [1; 2; 1; 2; 1; 2; ...]. If we call [clean_until] from two threads: 
* A wants to delete until value is 1 
* B want to delete until value is 2

We'd expect at most 3 items to be removed, depending whether thread A or B linearize first. But it seems that they could delete more than that if thread A happens to be running when 2 is the value of `head`, and thread B happens to be running when `1` is the `head`.

Reply:
DSCheck: 

```
 let test () = 
  let queue = create () in 
  push queue 1; 
  push queue 2;
  push queue 1; 
  push queue 2;
  push queue 1; 
  push queue 2;
  push queue 1; 
  push queue 2;
  Atomic.spawn (fun () -> 
    clean_until queue (fun v -> v == 1));
  Atomic.spawn (fun () -> 
    clean_until queue (fun v -> v == 2));
  Atomic.final (fun () -> 
    Atomic.check (fun () -> 
      not (is_empty queue)))
```